### ISSUE TITLE
[Bug]: Create Proxy with a Non-object as Target or Handler

### ISSUE THREAD

**Issue by clhiker** [babel/babel#14413]
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

@babel/cli

### Input code

```js
(async function () {
    function v0(v1 = () => null) {
    }
    await null;
});
const v0 = { alpha: { beta: { charlie: { delta: 'hello' } } } };
for (let v0 = 0; v0 < [].length; ++v0) {
    if (v0 > 8) {
        break;
    }
}
const v3 = new Proxy(v0, {
    has: function (v3, v4) {
        return v4 === 'value' ? false : Reflect.has(v3, v4);
    }
});
```
Here is the [PERL](https://babel.dev/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=false&spec=true&loose=true&code_lz=BQQwzgngdgxgBAMwK6wC4EsD2U7AJRwDeAUHGYijBtnAG4AMwtAjHALy4FsB8cUSAGwEES5OAF9S5EAHcQ6VH0ECA3MXF41MbGEUN2ROCAEAHABYgAXIYBGAU1RXDMCwCcB6O9cJwAJnYFHawByMwCBTGCJaPEJNQRMV1wBBzp6A3oVNLgAHjgAbQBdADoUqABzVDMsgGoahhEpMnQEXH1eAA5GsTEbVzsQAGs1MUlJbShdOgBmAyg7GTgABVdMAA8IJnoAGiImuAswa2Q0LBwmad3aABZunrJ-1CRXHBv2Ng5g2mMkOyiAfkQxjAdjg1gASnYECkqMVDhcrrcRuQxpogA&debug=false&forceAllTransforms=true&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=unambiguous&lineWrap=true&presets=env&prettier=false&targets=&version=7.17.8&externalPlugins=&assumptions=%7B%7D)

### Configuration file name

babel.config.json

### Configuration

```json
{
  "sourceType": "unambiguous",
  "presets": ["@babel/preset-env"],
  "plugins": [
    "@babel/plugin-transform-runtime"
  ]
}
```

### Current and expected behavior

The transpiler code run by node report an error:
```shell
var v3 = new Proxy(v0, {
         ^

TypeError: Cannot create proxy with a non-object as target or handler
```
This may be caused by Babel converting let variable to var. in the following new proxy, there is a bug caused by scope change

### Environment

- babel 
- - ├── @babel/cli@7.17.0=6
- - ├── @babel/generator@7.17.3
- - ├── @babel/plugin-transform-runtime@7.17.0
- - ├── @babel/preset-env@7.16.11
- Node: [v17.3.0]
- npm version [8.5.2]
- os [ubuntu 20.04]

**Comment by nicolo-ribaudo** [comment]
A smaller reproduction is
```js
let a = 0;
for (let a = 1;;) break;
console.log(a);
```

**Comment by nicolo-ribaudo** [comment]
This might be a bug in , which modifies the scope in a wrong way. This is the minimal code to see the bug:
```js
(function* () { var v0; });

const v0 = 1;
for (let v0 = 2;;) break;

console.log(v0);
```
With just the `@babel/plugin-transform-block-scoping` and `@babel/plugin-transform-regenerator` plugins.

**Comment by nicolo-ribaudo** [comment]
Yup, this is a regenerator bug. This is the minimal bug reproduction:
```js
const code = `
(function* () { var v0; });

var v0;
`;

const out = babel.transformSync(code, {
  configFile: false,
  plugins: [
    require("regenerator-transform"),
    () => {
      return {
        visitor: {
          Program: {
            enter(path) {
              console.log(Object.keys(path.scope.bindings));
            },
            exit(path) {
              console.log(Object.keys(path.scope.bindings));
            },
          },
        },
      };
    },
  ],
});
```

It should log `["v0"]` both times.

**Comment by nicolo-ribaudo** [comment]
Actually, it would be better to fix this directly in `path.traverse()`. Consider this example:
```js
const code = `
function f() { }
var v = 1;
`;

const out = babel.transformSync(code, {
  configFile: false,
  plugins: [
    ({ template }) => {
      return {
        visitor: {
          Function(path) {
            path
              .get("body")
              .pushContainer("body", template.statement.ast`var v = 0`);

            // path.scope.crawl();
          },
          VariableDeclaration(path) {
            if (path.node.declarations[0].init.value === 0) path.remove();
          },
          Program: {
            enter(path) {
              console.log(Object.keys(path.scope.bindings));
            },
            exit(path) {
              console.log(Object.keys(path.scope.bindings));
            },
          },
        },
      };
    },
  ],
});
```

It transforms the code like this:
<table>
<thead>
<tr>
  <th></th>
  <th>Initial</th>
  <th>After <code>Function</code> visitor</th>
  <th>After <code>VariableDeclaration</code> visitor on <code>var v = 0;</code></th>
</tr></thead>
<tr><td>Code</td><td>

```js
function f() { }
var v = 1;
```

</td><td>

```js
function f() { var v = 0; }
var v = 1;
```

</td><td>

```js
function f() {  }
var v = 1;
```

</td></tr>
<tr>
  <td>Global scope</td>
  <td><code>f</code>, <code>v</code></td>
  <td><code>f</code>, <code>v</code></td>
  <td><code>f</code></td>
</tr><tr>
  <td>Function scope</td>
  <td>empty</td>
  <td>empty</td>
  <td>empty</td>
</tr><tr>
  <td>Notes</td>
  <td></td>
  <td>The visitor didn't register the new <code>v</code> binding)</td>
  <td><code>path.remove()</code> wanted to remove the <code>v</code> binding of <code>var v = 0</code>, but since it wasn't registered it "leaked" to the other binding</td>
</tr></table>

We can fix this in https://github.com/babel/babel/blob/954177842c0082447b47d9c77a44a7df6e4df8f6/packages/babel-traverse/src/path/removal.ts#L25-L28.

Instead of removing all the binding identifiers with `removeBinding`, we should do something like this:
1. Detect if we need to use `var` semantics:
    1. If the current node is a VariableDeclaration,
       1. `isVar = node.kind === "var"`
    2. else, if the current node is a `FunctionDeclaration`, `ClassDeclaration` or `ImportDeclaration`,
       1. `isVar = false`
    3. else, if the parent node is an `ObjectProperty` whose parent is an `ObjectPattern` (and the current node is the `ObjectProperty`'s `.value`),
       1. repeat from 1.i using the `ObjectPattern` instead of the current node
    5. else if the parent node is an `AssignmentPattern` (and the current node is its `.left`), or an `ArrayPattern`, or a `VariableDeclarator` (and the current node is its `.id`),
       1. repeat from 1.i using the parent instead of the current node
    7. else if the parent node is an `ModuleSpecifier` or an `ImportDeclaration`,
       1. repeat from 1.i using the parent instead of the current node
    6. else,
      we are not in a declaration, we can completely exit from the function without deleting any binding
1. Use `getBindingIdentifiers` to get the list of bindings we should remove
1. If `isVar` is true remove them with `.removeOwnBinding` on `scope.getFunctionParent() || scope.getProgramParent()`, else remove them with `scope.removeOwnBinding()`

This should make sure that we only remove the correct bindings, and would also fix cases like this one:
```js
const code = `
var a = 1;
[a] = [];
`;

const out = babel.transformSync(code, {
  configFile: false,
  plugins: [
    () => {
      return {
        visitor: {
          ArrayPattern(path) {
            path.get("elements.0").remove();
          },
          Program: {
            enter(path) {
              console.log(Object.keys(path.scope.bindings));
            },
            exit(path) {
              console.log(Object.keys(path.scope.bindings)); // It should log [a], but it logs []
            },
          },
        },
      };
    },
  ],
});
```

---

@clhiker I'm assigning this to you since you selected "Would you like to work on a fix?". If you want to work on this please leave a comment!



If it is the first time that you contribute to Babel, follow these steps: (you need to have `make` and `yarn` available on your machine)
1. **Write a comment there** to let other possible contributors know that you are working on this bug.
1. Fork the repo
1. Run `git clone https://github.com/<YOUR_USERNAME>/babel.git && cd babel`
1. Run `yarn && make bootstrap`
1. Wait :hourglass_flowing_sand: 
1. Run `make watch` (or `make build` whenever you change a file)
1. Add a some tests in https://github.com/babel/babel/blob/main/packages/babel-traverse/test/removal.js (at least my first example in this issue, and one for every node handled by the above algorithm)
1. Update the code!
1. `yarn jest babel-traverse` to run the tests
1. If it is working, run `make test` to run all the tests
1. Run `git push` and open a PR!
